### ISSUE TITLE
Emit a Diagnostic if trying to emit a node from an unknown source. [phab:D1942]

### ISSUE THREAD

**Issue by kythe** [google/kythe#2389]
_Created by **benjyw** at 2017-09-22 00:09:51_: 

Previously, a NAME node would be emitted in such cases.
However NAME nodes are no longer part of the Kythe schema, so this
is needlessly confusing. 

This slightly simplifies the node building code, and removes
a source of confusion.

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2017-09-22 16:11:12_:

We like to favor continuing the analysis as much as possible instead of hard errors.  In that regard, emitting a Diagnostic would be better here.  Returning it like the name nodes were would be fine.  I have a TODO to have the rest of the indexer treat Diagnostics like null objects so no edges are pointed toward them.

**Comment by kythe** [comment]
_**schroederc** required changes at 2017-09-22 16:11:12_

**Comment by kythe** [comment]
_**benjyw** wrote at 2017-09-22 17:57:52_:

Makes sense. I remembered there being TODO nodes for this kind of thing, but they seem to have gone away. Didn't know about Diagnostic nodes. 

Done - although I had to slightly change the diagnostic emitting code to allow there to be no file context, since that's the very problem we're diagnosing...

This led me to notice that there seem to be quite a few places in the code that assume that getFileVName will never return null. So in practice we will get NPEs if this occurs. But fixing that is a project for another day, I reckon.

**Comment by kythe** [comment]
_**benjyw** wrote at 2017-09-22 17:58:58_:

    Address code review comments:
    
    - Emit a Diagnostic node instead of thrown a runtime exception.
    - Made it possible to emit Diagnostics not attached to any file,
      since the problem we're diagnosing is exactly that of not knowing
      the file context...

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2017-09-22 18:41:14_:

Use `warningfmt` instead of separately formatting the string.

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2017-09-22 18:41:14_:

`enclClass` can also be `null`.  Just log the `Symbol` and `ClassSymbol` (which will show the `null` if that's the case).

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2017-09-22 18:41:14_:

Instead of nesting, it might be better to do a quick return:

```
if (fileVName == null) {
  return dn;
} else if (d.hasSpan()) {
// ... existing code
```

**Comment by kythe** [comment]
_**schroederc** wrote at 2017-09-22 18:41:14_:

Please run all touched files through `google-java-format`.

Please update the change description.

Also, do you have an example when this case is hit?  If so, please file a bug.

**Comment by kythe** [comment]
_**schroederc** required changes at 2017-09-22 18:41:14_

**Comment by kythe** [comment]
_**benjyw** commented inline (context lost during migration) at 2017-09-22 20:31:41_:

Done, but for the record I'm not sure I concur, as the existing code is still nested at the same depth, but now there are two identical return statements. To me this seems less readable, buty YMMV.

I could also lose the else block, since the other branch always returns anyway, but some people don't like that as it's more bug-prone if the other branch is changed in the future.

**Comment by kythe** [comment]
_**benjyw** commented inline (context lost during migration) at 2017-09-22 20:31:41_:

I pass the message into the Diagnostic, which is why I format it separately.

**Comment by kythe** [comment]
_**benjyw** wrote at 2017-09-22 20:31:41_:

Thanks for the quick turnaround! See my responses inline.

I know of no case that actually triggers this, BTW.

**Comment by kythe** [comment]
_**schroederc** accepted this revision at 2017-09-22 21:14:46_

**Comment by kythe** [comment]
_**schroederc** closed via commit a7f385ac74d58c99fda1129d82bc7c7d5afa6594 at 2017-09-22 21:15:13_

**Comment by kythe** [comment]
_**benjyw** wrote at 2017-09-22 23:49:43_:

Thanks as always for the quick turnaround!
